runtime/lua/vim/pack.lua

```diff
@@ -0,0 +1,911 @@
+    return
+  end
+
+  n_added_plugins = n_added_plugins + 1
+  added_plugins[plug.path] = { plug = plug, id = n_added_plugins }
+
+  vim.cmd.packadd({ plug.spec.name, bang = bang })
+
+  -- Execute 'after/' scripts if not during startup (when they will be sourced
+  -- automatically), as `:packadd` only sources plain 'plugin/' files.
+  -- See https://github.com/vim/vim/issues/15584
+  -- Deliberately do so after executing all currently known 'plugin/' files.
+  local should_load_after_dir = vim.v.vim_did_enter == 1 and not bang and vim.o.loadplugins
+  if should_load_after_dir then
+    local after_paths = vim.fn.glob(plug.path .. '/after/plugin/**/*.{vim,lua}', false, true)
+    --- @param path string
+    vim.tbl_map(function(path)
+      pcall(vim.cmd.source, vim.fn.fnameescape(path))
+    end, after_paths)
+  end
+end
+
+--- @class vim.pack.keyset.add
+--- @inlinedoc
+--- @field bang? boolean Whether to execute `:packadd!` instead of |:packadd|. Default `false`.
+
+--- Add plugin to current session
+---
+--- - For each specification check that plugin exists on disk in |vim.pack-directory|:
+---     - If exists, do nothin in this step.
+---     - If doesn't exist, install it by downloading from `source` into `name`
+---       subdirectory and update state to match `version`.
+--- - For each plugin execute |:packadd| making them reachable by Nvim.
+---
+--- Notes:
+--- - Installation is done in parallel, but waits for all to finish before
+---   continuing next code execution.
+--- - If plugin is already present on disk, there are no checks about its present state.
+---   The specified `version` can be not the one actually present on disk.
+---   Execute |vim.pack.update()| (possibly with `{ offline = true }`) to synchronize.
+--- - Adding plugin second and more times during single session does nothing:
+---   only the data from the first adding is registered.
+---
+--- @param specs (string|vim.pack.Spec)[] List of plugin specifications. String item
+--- is treated as `source`.
+--- @param opts? vim.pack.keyset.add
+function M.add(specs, opts)
+  vim.validate('specs', specs, vim.islist, false, 'list')
+  opts = vim.tbl_extend('force', { bang = false }, opts or {})
+  vim.validate('opts', opts, 'table')
+
+  --- @type vim.pack.Plug[]
+  local plugs = vim.tbl_map(new_plug, specs)
+  -- TODO(echasnovski): Normalize all plugins as a whole, mostly process
+  -- duplicates. This has little benefit now, but will be more sore after
+  -- packspec processing (as it might introduce conflicting dependencies).
+
+  -- Install
+  --- @param p vim.pack.Plug
+  local plugs_to_install = vim.tbl_filter(function(p)
+    return uv.fs_stat(p.path) == nil
+  end, plugs)
+  local pluglist_to_install = PlugList.new(plugs_to_install)
+  if #plugs_to_install > 0 then
+    git_ensure_exec()
+    pluglist_to_install:install()
+  end
+
+  -- Register and `:packadd` those actually on disk
+  for _, p in ipairs(plugs) do
+    if uv.fs_stat(p.path) ~= nil then
```

Comment:
This will only not add plugins that did not install to disk. Plugin can still be installed but not in correct version after installation (like if target doesn't exist), which will be shown as error during initial install and error during update.

This decision is mostly due to the fact that this plugin would still load after restart (as there is no check if the version is correct; for performance).